src/nekrs_driver.cpp

```diff
@@ -0,0 +1,229 @@
+  nekrs::copyToNek(time_, tstep_);
+}
+
+void NekRSDriver::write_step(int timestep, int iteration)
+{
+  nekrs::copyToNek(timestep, iteration);
+  nekrs::nekOutfld();
+  return;
+}
+
+Position NekRSDriver::centroid_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+  Position c{0., 0., 0.};
+  double mass = 0.;
+  for (gsl::index i = 0; i < n_gll_; ++i) {
+    auto idx = local_elem * n_gll_ + i;
+    c.x += x_[idx] * mass_matrix_[idx];
+    c.y += y_[idx] * mass_matrix_[idx];
+    c.z += z_[idx] * mass_matrix_[idx];
+    mass += mass_matrix_[idx];
+  }
+  c.x /= mass;
+  c.y /= mass;
+  c.z /= mass;
+  return c;
+}
+
+std::vector<Position> NekRSDriver::centroid_local() const
+{
+  std::vector<Position> c(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    c[i] = this->centroid_at(i);
+  }
+  return c;
+}
+
+double NekRSDriver::volume_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+  double v = 0.;
+  for (int32_t i = 0; i < n_gll_; ++i) {
+    v += mass_matrix_[local_elem * n_gll_ + i];
+  }
+  return v;
+}
+
+std::vector<double> NekRSDriver::volume_local() const
+{
+  std::vector<double> v(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    v[i] = this->volume_at(i);
+  }
+  return v;
+}
+
+double NekRSDriver::temperature_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+
+  double sum0 = 0.;
+  double sum1 = 0.;
+  for (int32_t i = 0; i < n_gll_; ++i) {
+    auto idx = local_elem * n_gll_ + i;
+    sum0 += rho_cp_[idx] * temperature_[idx];
+    sum1 += rho_cp_[idx];
+  }
+  return sum0 / sum1;
+}
+
+std::vector<double> NekRSDriver::temperature_local() const
+{
+  std::vector<double> t(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    t[i] = this->temperature_at(i);
+  }
+
+  return t;
+}
+
+std::vector<double> NekRSDriver::density_local() const
+{
+  nekrs::copyToNek(time_, tstep_);
+  std::vector<double> local_densities(n_local_elem());
+
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    if (this->in_fluid_at(i) == 1) {
+      auto T = this->temperature_at(i);
+      // nu1 returns specific volume in [m^3/kg]
+      local_densities[i] = 1.0e-3 / iapws::nu1(pressure_bc_, T);
+    } else {
+      local_densities[i] = 0.0;
+    }
+  }
+
+  return local_densities;
+}
+
+int NekRSDriver::in_fluid_at(int32_t local_elem) const
+{
+  // In NekRS, element_info_[i] == 1 if i is a *solid* element
+  Expects(local_elem < n_local_elem());
+  return element_info_[local_elem] == 1 ? 0 : 1;
+}
+
+std::vector<int> NekRSDriver::fluid_mask_local() const
+{
+  std::vector<int> mask(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    mask[i] = in_fluid_at(i);
+  }
+  return mask;
+}
+
+int NekRSDriver::set_heat_source_at(int32_t local_elem, double heat)
+{
+  for (int i = 0; i < n_gll_; ++i) {
+    localq_->at(local_elem * n_gll_ + i) = heat;
+  }
+  return 0;
+}
+
+void NekRSDriver::open_lib_udf()
```

Comment:
Here's what happening in this function:

1. At runtime, NekRS compiles `libUDF.so` from the `.udf` file and then dynamically loads it.  We're doing the same here.  `dlopen` works fine if it's called multiple times with the same library, since the dynamic library handle has a reference count (see the [man page for dlopen and dlclose](https://linux.die.net/man/3/dlopen)).  In the NekRSDriver header, the name `libUDF.so` has been hardcoded, which is not great; however, NekRS also hardcodes the library name as `libUDF.so`, so I figured it was a suitable to do the same.  

2. The `.udf` file is assumed to define global variable named `localq`, which is a `std::vector<double>` that contains the element-wise heat sources.  This is not a standard or required variable for a general `.udf` file, but I'm requiring it for our coupling implementation.  So, that symbol is dynamically loaded here.  I've tried to guard against unexpected behavior by checking `dlsym` errors and by accessing `localq_` via the `at()` method (see `NekRSDriver::set_heat_source_at`.

Reply:
Fixed in https://github.com/enrico-dev/enrico/pull/113/commits/411c99a6a9c8bcbe2af7450d8218d19c8d8140e2